shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1272,43 +1277,67 @@ - (void)updateViewportPadding {
-    return;
+    // Ignore keyboard notifications if engine’s viewController is not current viewController.
+    if ([_engine.get() viewController] != self) {
+      return;
+    }
   }
 
-  CGRect keyboardFrame = [[info objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
   CGRect screenRect = [[UIScreen mainScreen] bounds];
 
-  // Get the animation duration
+  // In Slide Over view, the keyboard's dimensions/position does not include the space
+  // below the app, even though the keyboard may be at the bottom of the screen.
+  // To handle, shift the Y origin by the amount of space below the app.
+  CGFloat screenHeight = CGRectGetHeight(screenRect);
+  CGFloat screenWidth = CGRectGetWidth(screenRect);
+  CGFloat appHeight = CGRectGetHeight(self.view.window.frame);
+  CGFloat appWidth = CGRectGetWidth(self.view.window.frame);
+  if (self.view.safeAreaInsets.bottom > 0 && appWidth < screenWidth) {
+    // In Slide Over view, the app is vertically centered with space above and below,
+    // which is why we divide by 2 to get the space below.
+    keyboardFrame.origin.y += (screenHeight - appHeight) / 2;
+  }
+
+  // If keyboard is within the screen and it's not empty, calculate and set the inset.
+  // If keyboard is not within the screen (it's usually below), set inset to 0.
+  // If keyboard frame is empty (usually because it was dragged and dropped), set inset to 0.
+  CGFloat calculatedInset = 0;
+  if (CGRectIntersectsRect(keyboardFrame, screenRect) && !isEmpty) {
+    calculatedInset = [self calculateKeyboardInset:screenRect keyboardFrame:keyboardFrame];
+  }
+
+  // avoid double triggering startKeyBoardAnimation
+  if (self.targetViewInsetBottom == calculatedInset) {
+    return;
+  }
+
+  self.targetViewInsetBottom = calculatedInset;
   NSTimeInterval duration =
       [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
-
-  // Considering the iPad's split keyboard, Flutter needs to check if the keyboard frame is present
-  // in the screen to see if the keyboard is visible.
-  if (CGRectIntersectsRect(keyboardFrame, screenRect)) {
-    CGFloat bottom = CGRectGetHeight(keyboardFrame);
-    CGFloat scale = [UIScreen mainScreen].scale;
-    // The keyboard is treated as an inset since we want to effectively reduce the window size by
-    // the keyboard height. The Dart side will compute a value accounting for the keyboard-consuming
-    // bottom padding.
-    self.targetViewInsetBottom = bottom * scale;
-  } else {
-    self.targetViewInsetBottom = 0;
-  }
   [self startKeyBoardAnimation:duration];
 }
 
-- (void)keyboardWillBeHidden:(NSNotification*)notification {
+- (void)keyboardWillChangeFrame:(NSNotification*)notification {
+  // Immediately prior to a change in keyboard frame, this notification is triggered.
+  // There are some cases where UIKeyboardWillShowNotification & UIKeyboardWillHideNotification
+  // do not act as expected and this is used to catch those cases.
```

Comment:
So none of them are very reliable which is why I used all three of them. Here's some of the use cases for example
* When predictive-only/minimized keyboard is dragged, it sends a `UIKeyboardWillShowNotification` notification where the keyboard = CGRectZero, which we use to determine it's "floating" and should have inset 0 . Whereas in `UIKeyboardWillChangeFrameNotification` we skip when keyboard = CGRectZero, because otherwise it will set inset to 0 prematurely in some cases like when you dragged a docked keyboard but don't actually change it from docked. 
* When keyboard is changed from docked to floating, it's supposed to send a `UIKeyboardWillHideNotification` notification but it doesn't, so we handle in `UIKeyboardWillChangeFrameNotification` instead
* When keyboard goes from docked to undocked, the `UIKeyboardWillChangeFrameNotification` position will not be completely below the screen so it will still have an inset, which is at least one reason why we also need `UIKeyboardWillHideNotification`

Reply:
Ah, gotcha. I think logic this complex deserves a design documentation :) And we should have a comment in the code to link to the design doc. (The same way PlatformView should have and never did :p )

I think for this PR, it would be nice to explain "some cases" a little more in the comments. Or add a link in the code to your comment here: https://github.com/flutter/engine/pull/37719#discussion_r1035437562